contrib/clients/xcpustate/xcpustate.man
.TH XCPUSTATE 1 "13 July 1988" "X Version 11"
.SH NAME
xcpustate - display CPU states (idle, nice, system, kernel) statistics
.SH SYNTAX
\fBxcpustate\fP [\fI-toolkitoption\fP ...] [\fI-count\fP iterations] 
[\fI-interval\fP seconds]
.SH DESCRIPTION
.I Xcpustate
displays bars showing the percentage of time the CPU spends in
different states. 
.PP
On a machine running Berkeley Unix, or a derivative (Suns with SunOS,
microVaxen with Ultrix), the bar indicates the proportions of idle,
user, nice, and system time with increasing levels of gray scale (from
left to right) 
.PP
On an Iris4D, there may be more than one bar, one for each CPU, and
each bar indicates the proportions of idle + wait, user, kernel, sxbrk
and interrupt time for each cpu.
.SH OPTIONS
.I Xcpustate
accepts all of the standard X Toolkit command line options, plus:
.TP 8
.IR -count " iterations"
The number of times it should update the display. Default is forever.
.TP 8
.IR -interval " seconds"
the interval in between updates. Default is 1 second.
.SH X DEFAULTS
For
.I xcpustate
the available class identifiers are:
.sp
.nf
CPUStateMonitor - the application
Form - enclosing the entire application, and sub-Forms enclosing 
Label/Bar pairs.
.fi
.PP
For
.I xcpustate,
the available name identifiers are:
.sp
.nf
xcpustate - application name
The outer Form is "form".
The Forms enclosing the Label/Bar pairs are "formN", where N is the
index number, starting with the top pair as zero.
Each Label name is the same as the label string.
Each Bar name is "barN".
.fi
.sp
.LP
For
.I xcpustate,
the available resources are:
.IP "name interval, class Interval"
corresponds to the -interval option. Takes an integer value.
.IP "name count, class Count"
corresponds to the -count argument, also takes and integer value.
.SH NOTES
.I Xcpustate 
was designed for a multiprocessor (a Silicon Graphics Iris4D/240) - it
happens to work for conventional BSD uniprocessors. It is meant to be
easy to port, and extend to monitor a wide variety of statistics.
.SH SEE ALSO
xperfmon, xload
.SH AUTHOR
Mark Moraes, University of Toronto
(Thanks to Chris Siebenmann for the code for BSD systems)


contrib/clients/xcpustate/xcpustate.c
/*
*  Displays CPU state distribution
*/
#ifndef lint
static char rcsid[] = "$Header: /csri3/X.V11R3/localstuff/xcpustate/RCS/xcpustate.c,v 1.2 89/10/05 23:03:11 xwindows Exp $";
#endif /*lint*/

#include <stdio.h>
#include <sys/types.h>

#include <X11/StringDefs.h>
#include <X11/Intrinsic.h>
#include <X11/Form.h>
#include <X11/Label.h>
#include <X11/Xmu.h>
#include "Bar.h"
#include "s.h"

#define MAXSTR 512
#define DEF_INTERVAL	1
#define DEF_COUNT		-1

char *progname;
static int defaultInterval = DEF_INTERVAL;
static int defaultCount = DEF_COUNT;
static int count, interval;

/* Application Resources - no particular widget */
static XtResource application_resources[] = {
    {"interval", "Interval", XtRInt, sizeof(int),
	    (Cardinal)&interval, XtRInt, (caddr_t) &defaultInterval},
    {"count", "Count", XtRInt, sizeof(int),
	    (Cardinal)&count, XtRInt, (caddr_t) &defaultCount},
};

/*
*  Command line options table. The command line is parsed for these,
*  and it sets/overrides the appropriate values in the resource
*  database
*/
static XrmOptionDescRec optionDescList[] = {
    {"-interval",   ".interval",    XrmoptionSepArg,    (caddr_t) NULL},
    {"-count",	    ".count",	    XrmoptionSepArg,    (caddr_t) NULL},
};

/*
*  DON'T CHANGE THE ORDER OF THE ARGS IN THE VARIOUS ARG STRUCTS. IF
*  YOU WANT TO ADD STUFF, ADD IT AT THE END OF THE STRUCT, BECAUSE WE
*  REFER TO SOME OF THE ELEMENTS BY POSITION IN THE CODE.
*/
/* No spacing between the widgets on the Form */
static Arg form_args[] = {
    {XtNdefaultDistance, (XtArgVal) 0},
};

static Arg subform_args[] = {
    {XtNfromVert, (XtArgVal) 0},
    {XtNdefaultDistance, (XtArgVal) 0},
    {XtNborderWidth, (XtArgVal) 0},
    {XtNtop, (XtArgVal) XtChainTop},
    {XtNbottom, (XtArgVal) XtChainTop}, /* ChainBottom causes strange resize */
    {XtNright, (XtArgVal) XtChainRight},
    {XtNleft, (XtArgVal) XtChainLeft},
};

static Arg bar_args[] = {
    {XtNfromHoriz, (XtArgVal) 0},
    {XtNorientation, (XtArgVal) XtorientHorizontal},
    {XtNborderWidth, (XtArgVal) 1},
    {XtNlength, (XtArgVal) ((Dimension) 200)},
    {XtNthickness, (XtArgVal) ((Dimension) 20)},
    {XtNtop, (XtArgVal) XtChainTop},
    {XtNbottom, (XtArgVal) XtChainTop},
    {XtNright, (XtArgVal) XtChainRight},
    {XtNleft, (XtArgVal) XtChainLeft},
    {XtNvertDistance, (XtArgVal) 0},
    {XtNhorizDistance, (XtArgVal) 0},
    {XtNresizable, (XtArgVal) TRUE},
};

static Arg label_args[] = {
    {XtNlabel, (XtArgVal) 0},
    {XtNjustify, (XtArgVal) XtJustifyLeft},
    {XtNborderWidth, (XtArgVal) 0},
    {XtNtop, (XtArgVal) XtChainTop},
    {XtNbottom, (XtArgVal) XtChainTop},
    {XtNright, (XtArgVal) XtChainLeft},
    {XtNleft, (XtArgVal) XtChainLeft},
    {XtNvertDistance, (XtArgVal) 0},
    {XtNhorizDistance, (XtArgVal) 0},
    {XtNresizable, (XtArgVal) FALSE},
};

void
usage()
{
    (void) fprintf(stderr, "\
Usage: %s [Xt options] [-count iterations] [-interval delay_seconds\n",
     progname);
    exit(-1);
}

char *
xmalloc(n)
int n;
{
    extern char *malloc();
    char *cp = malloc((unsigned) n);
    if (cp == NULL) {
	    (void) fprintf(stderr, "Couldn't malloc %d bytes\n", n);
	    exit(-1);
    }
    return cp;
}

static int nbars;
static Widget *bar;
static char **barnames;
    
main(argc, argv)
int argc;
char **argv;
{
    int i;
    static XtTimerCallbackProc update_display();
    Widget topLevel;
    Widget form, label;
	Widget subform = NULL;
    extern char *strchr(/* const char *, char */);

    if ((progname = strchr(argv[0], '/')) == NULL)
	progname = argv[0];
    else
	progname++;

    nbars = num_bars();
    if (nbars == 0) {
	(void) fprintf(stderr, "num_bars returned 0 - something is wrong\n");
	exit(-1);
    }
    bar = (Widget *) xmalloc(nbars * sizeof(Widget));
    barnames = label_bars(nbars);

    topLevel = XtInitialize(progname, "CPUStateMonitor",
			    optionDescList, XtNumber(optionDescList),
			    &argc, argv);

    if (argc > 1)
	usage();
	
    XtGetApplicationResources(topLevel, 0, application_resources,
			      XtNumber(application_resources), NULL, 0 );

    form = XtCreateManagedWidget("form", formWidgetClass, topLevel,
				 form_args, XtNumber(form_args));
    
    for(i = 0; i < nbars; i++) {
#define FORMNAMEFMT	"form%d"
	char formname[sizeof(FORMNAMEFMT) + 32];
#define BARNAMEFMT	"bar%d"
	char barname[sizeof(BARNAMEFMT) + 32];

	if (i > 0)
	    subform_args[0].value = (XtArgVal) subform;
	(void) sprintf(formname, FORMNAMEFMT, i);
	subform = XtCreateManagedWidget(formname, formWidgetClass, form,
				 subform_args, XtNumber(subform_args));
	label_args[0].value = (XtArgVal) barnames[i];
	label = XtCreateManagedWidget(barnames[i], labelWidgetClass, subform,
				      label_args, XtNumber(label_args));
	(void) sprintf(barname, BARNAMEFMT, i);
	bar_args[0].value = (XtArgVal) label;
	bar[i] = XtCreateManagedWidget(barname, barWidgetClass, subform,
				    bar_args, XtNumber(bar_args));
    }
    XtRealizeWidget(topLevel);

    init_bars(nbars);
    
    (void) XtAddTimeOut((unsigned long) (interval * 1000), update_display,
			(caddr_t) NULL);
    XtMainLoop();
}

/* ARGSUSED */
static XtTimerCallbackProc
update_display(closure, id)
caddr_t closure;
XtIntervalId *id;
{
    display_bars(nbars);
    
    if (count > 0) {
	if (--count == 0)
	    return;
    }
    (void) XtAddTimeOut((unsigned long) (interval * 1000), update_display,
			(caddr_t) NULL);
}

void
draw_bar(i, states, nstates)
int i;
int states[];
int nstates;
{
    SetBarValues((Widget) bar[i], states, nstates);
}
